xapian-applications/omega/handler_libarchive.cc

```diff
@@ -0,0 +1,122 @@
+/** @file handler_libarchive.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "metaxmlparse.h"
+#include "opendocparse.h"
+#include "str.h"
+
+#include<archive.h>
+#include<archive_entry.h>
+
+#include <cstring>
+
+using namespace std;
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	struct archive* archive_obj;
+	struct archive_entry* entry;
+	int status_code;
+	const char* file = filename.c_str();
+	archive_obj = archive_read_new();
+	archive_read_support_filter_all(archive_obj);
+	archive_read_support_format_all(archive_obj);
+	status_code = archive_read_open_filename(archive_obj, file, 10240);
+
+	if (status_code != ARCHIVE_OK) {
+	    error = "Libarchive failed to open the file " + filename;
+		return false;
+	}
+
+	size_t total;
+	ssize_t size;
+	string s;
+
+	// extracting data from content.xml and styles.xml
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    if (strcmp(archive_entry_pathname(entry), "content.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf1[total];
```

Comment:
This is not a proper way to declare variable length arrays in C++ as some version of the standard doesn't support it. 
You can use a constant-expression or the operators `new` and `delete`.

Reply:
Yes - ISO C99 added VLAs, and some C++ compilers support them in C++ code as a non-standard extension, but we shouldn't rely on that as it limits portability.

The size needs to be a compile-time constant - you could allocate a few KB and then call `archive_read_data()` repeatedly to read the file in chunks.

If you want a variable size block, the portable C++ version is something like:

```
unique_ptr<char> buf1 = new char[total];
```

Though here you can avoid allocating a temporary buffer entirely by resizing the string to have enough space and then reading directly into that space, which also avoids an extra copy operation:

```
auto i = s.size();
s.resize(i + total);
size = archive_read_data(archive_obj, &s[i], total);
if (size <= 0) {
    // ...
}
s.resize(i + size);

Reply:
Also, `buf1`, `buf2`, `buf3` aren't great variable names unless the numbers have some actual meaning (which they don't here).  You could actually just call them all `buf` in this case I think, since they each exist in disjoint blocks, but if they were in the same block then `buf_content`, `buf_styles`, `buf_meta` would be easier for the reader to follow than arbitrary numbers.

Reply:
I have used string resizing now. Thanks for pointing it out, I have renamed the buffers to `buf`. As they are in different blocks, it shouln't be a problem.